### ISSUE TITLE
Documentation for the cluster protocol

### ISSUE THREAD

**Issue by delfick** [nats-io/gnatsd#334]
This is a transferal of https://github.com/nats-io/nats/issues/215, which should have been created here.

---------

Hi,

We have a situation where we have gnats instances in a mesos setup, which means gnats instances can be killed and moved to somewhere else in the cluster at random times (i.e. a mesos slave going down).

We are currently running our gnats such that the --cluster is the same port as the -p option and the --routes points at the port our load balancer serves for the cluster of gnats.

This tends to work fine, however we’ve noticed that with enough traffic the clustering seems to fall apart and not recover.

To workaround this issue (for at least just the short term) we think some integration with mesos would be nice, where the --routes points at something that talks to mesos and gives back information about the active/dead gnats instances as they move around.

I tried making such a bridge today but I’m having a difficulty trying to work out a) if it’s even possible and b) what the protocol actually is.

So, long story short, if our idea is even feasible, can we get some documentation on what would be involved in implementing the clustering protocol?

Thanks!

**Comment by kozlovic** [comment]
I would really recommend using different ports for the client and route connections. Since by default the server will listen to any interface (if you don't specify `-a`, it listens to `0.0.0.0`) for the client connections. For the route, it will be what you specify in `-cluster`.
If the other server that tries to connect is given an interface that is not the one of the cluster, but one that `0.0.0.0` resolves to on that machine, your server will create a connection to the client port, which is handled differently. That's not good. Having different ports ensure that this is not going to happen.

In the original issue, you have reported that you are using `0.8.1`. I would strongly recommend upgrading to our latest `0.9.4` and with the above in mind, do more tests and report back to us to see if the problems persist.

Thanks!

**Comment by delfick** [comment]
Ok, so the good news is my eyes need checking and we actually are already using different ports in production for --cluster and -p. So that's actually not a problem we have :)

We've also decided to set up a situation where we have 1 seed gnats that sits there purely listening  on a --cluster port and several non-seed gnats that do the actual work.

Where the seed gnats has a --route pointing at the load balancer to find one of the non-seed servers and the non-seed servers have a --route pointing at the seed server. Hopefully this solves the problem where the --route points at itself through the load balancer, and should solve the split brain because the seed server will be connected to by all the non-seed servers and then seed server should spread the word back to them about everything that has connected to it.

We'll also upgrade gnatsd itself at some point.

Thanks.

**Comment by kozlovic** [comment]
From what you wrote, it sounds like the non-seed servers access to the seed IP? If so, then this should be enough. Note that you could have more than one seed server for HA. The non-seed servers could have `-routes nats://seed1:port,nats://seed2:port`.
I guess that you mean that the seed could try to connect to one of the non-seed (with url pointing to load balancer) so that if the seed gets separated from the reset (network partition), it will then reconnect to one of the non-seed server and the mesh would reform. Is that what it is?
Is the load-balancer potentially returning the seed IP? That would be a problem: not that it is an error for a server to point a route to itself, it is detected as a route to self and ignored, but it would prevent the reconnect to the mesh.

**Comment by ColinSullivan1** [comment]
Documentation of the protocol is underway.

**Comment by delfick** [comment]
awesome, thanks!

**Comment by ColinSullivan1** [comment]
Resolved with https://github.com/nats-io/nats-site/pull/151.

**Comment by delfick** [comment]
Excellent, thank you!

I'm also happy to report that gnats under 0.9.4 and the new configuration I mentioned before (one seed server seperate to the cluster that does the work) is a lot more stable than before :)

**Comment by kozlovic** [comment]
@delfick Thank you for the confirmation, and the good news 😄
